lib/DependencyProvider.js

```diff
@@ -0,0 +1,350 @@
+  const { pkg, version } = splitPkgVersion(newVersions.pop());
+  const cachePkgVersions = onlineVersionsCache.get(pkg);
+  if (
+    cachePkgVersions !== undefined &&
+    cachePkgVersions[cachePkgVersions.length - 1] === version
+  ) {
+    // Insert (in reverse) newVersions into onlineVersionsCache.
+    for (const pkgVersion of newVersions.reverse()) {
+      const { pkg, version } = splitPkgVersion(pkgVersion);
+      const versionsOfPkg = onlineVersionsCache.get(pkg);
+      if (versionsOfPkg === undefined) {
+        onlineVersionsCache.set(pkg, [version]);
+      } else {
+        versionsOfPkg.push(version);
+      }
+    }
+    // Save the updated onlineVersionsCache to disk.
+    const onlineVersions = fromEntries(onlineVersionsCache.entries());
+    fs.writeFileSync(cachePath, JSON.stringify(onlineVersions));
+  } else {
+    // There was a problem and a package got deleted from the server.
+    updateCacheFromScratch(cachePath, remotePackagesUrl);
+  }
+}
+
+function listAvailableVersionsOnline(pkg /*: string */) /*: Array<string> */ {
+  const memoVersions = listVersionsMemoCache.get(pkg);
+  if (memoVersions !== undefined) {
+    return memoVersions;
+  }
+  const offlineVersions = listAvailableVersionsOffline(pkg);
+  const allVersionsSet = new Set(versionsFromOnlineCache(pkg));
+  // Combine local and online versions.
+  for (const version of offlineVersions) {
+    allVersionsSet.add(version);
+  }
+  const allVersions = [...allVersionsSet].sort().reverse();
+  listVersionsMemoCache.set(pkg, allVersions);
+  return allVersions;
+}
+
+// onlineVersionsCache is a Map with pkg as keys.
+function versionsFromOnlineCache(pkg /*: string */) /*: Array<string> */ {
+  const versions = onlineVersionsCache.get(pkg);
+  return versions === undefined ? [] : versions;
+}
+
+function listAvailableVersionsOffline(pkg /*: string */) /*: Array<string> */ {
+  const memoVersions = listVersionsMemoCache.get(pkg);
+  if (memoVersions !== undefined) {
+    return memoVersions;
+  }
+
+  // console.log('List versions of: ' + pkg);
+  let offlineVersions;
+  try {
+    offlineVersions = fs.readdirSync(homePkgPath(pkg));
+  } catch (_) {
+    // console.log(
+    //   `Directory "${homePkgPath(pkg)}" does not exist for package ${pkg}.`
+    // );
+    // console.log(
+    //   `Offline mode, so we return [] for the list of versions of ${pkg}.`
+    // );
+    offlineVersions = [];
+  }
+
+  // Reverse order of subdirectories to have newest versions first.
+  offlineVersions.reverse();
+  listVersionsMemoCache.set(pkg, offlineVersions);
+  return offlineVersions;
+}
+
+// Helper functions ##################################################
+
+// We can replace this with using `Object.fromEntires` once Node.js 10 is
+// EOL 2021-04-30 and support for Node.js 10 is dropped.
+function fromEntries(entries) {
+  const res = {};
+  for (const [key, value] of entries) {
+    res[key] = value;
+  }
+  return res;
+}
```

Comment:
This should be removed now and we can use `Object.fromEntries()` instead since has been merged

Reply:
I'll remove it thanks!